File: src/_5_medium_longest_palindromic_substring.ts
import { assertEquals } from "test";
/*
 * @lc app=leetcode id=5 lang=typescript
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
function longestPalindrome(s: string): string {
  if (s.length < 2) {
    return s;
  }

  let start = 0;
  let maxLength = 1;
  let i = 0;

  while (i < s.length) {
    let left = i;
    let right = i;

    while (right < s.length - 1 && s.at(right) === s.at(right + 1)) {
      right++;
    }
    i = right + 1;
    while (
      right < s.length - 1 && left > 0 && s.at(right + 1) === s.at(left - 1)
    ) {
      left--;
      right++;
    }
    if (maxLength < right - left + 1) {
      start = left;
      maxLength = right - left + 1;
    }
  }
  return s.slice(start, start + maxLength);
}
// @lc code=end

Deno.test("longestPalindrome", (_) => {
  assertEquals(longestPalindrome("abacd"), "aba");
});

File: src/_5_medium_longest_palindromic_substring.rs
/*
 * @lc app=leetcode id=5 lang=rust
 *
 * [5] Longest Palindromic Substring
 */

struct Solution {}

// @lc code=start
impl Solution {
    pub fn longest_palindrome(s: String) -> String {
        if s.len() < 2 {
            return s;
        }
        let mut start = 0;
        let mut max_len = 1;
        let mut i = 0;

        let chars = s.chars().collect::<Vec<char>>();

        while i < s.len() {
            let mut left = i;
            let mut right = i;

            while right < s.len() - 1 && chars.get(right) == chars.get(right + 1) {
                right += 1;
            }

            i = right + 1;

            while right < s.len() - 1 && left > 0 && chars.get(right + 1) == chars.get(left - 1) {
                left -= 1;
                right += 1;
            }

            if max_len < right - left + 1 {
                start = left;
                max_len = right - left + 1;
            }
        }

        String::from(&s[start..start + max_len])
    }
}
// @lc code=end

#[test]
fn test5() {
    assert_eq!(
        Solution::longest_palindrome(String::from("babad")),
        String::from("bab")
    );
    assert_eq!(
        Solution::longest_palindrome(String::from("cbbd")),
        String::from("bb")
    );
}
